### PR TITLE
Create rapid_observations.toml

### PR DESCRIPTION
**Pull request by ekamau**
Initial draft of observations table for the RAPID study.

### CODE REVIEW

**Comment by pipliggins** [comment]
Could you create this branch off 'rapid' rather than 'main', as that's where we're developing the rapid parser? (although it's probably complete enough to start merging into the main branch now that they're both at similar stages of development)

**Comment by ekamau** [comment]
I could try .. can the branch be merged with 'rapid' or there is no such
thing in git? or move the rapid-observations.toml to the branch 'rapid'? ..

**Comment by pipliggins** [comment]
I think because I recently merged `main` into `rapid` so they should be up to date with each other, you should be able to do this... I've never tried it though so not 100% sure what it will do!
```
git checkout rapid-observations # Just to be clear which branch to be on.
git rebase --onto rapid main rapid-observations
```

Alternatively if you've only got the one file, you could save that externally, then delete this branch and recreate it using 'rapid' as the starting point using `git checkout -b rapid-observations rapid`, add the file there then recreate the pull request.

**Comment by ekamau** [comment]
Fantastic. Thanks.

**Comment by abhidg** [comment]
GitHub also allows you to change the base branch from the web interface :)

File: isaric/parsers/rapid_observations.toml [hunk 1]
```diff
@@ -0,0 +1,782 @@
...
+{ field = "dsstdtc", description = "Outcome date" }
+]
+
+[[observation]]
+name = "haemorrhage"
+date = { ref = "admissionDateHierarchy" }
+phase = "admission"
+is_present = { field = "bleed_ceoccur_v2", ref = "Y/N/NK" }
+if = { bleed_ceoccur_v2 = { "!=" = 3 } }
+occurrence_period =
```
Comment (abhidg) [inline isaric/parsers/rapid_observations.toml:24]: keys in TOML can't be left blank, you can comment it out meanwhile

File: isaric/parsers/rapid_observations.toml [hunk 2]
```diff
@@ -0,0 +1,782 @@
...
+is_present = { field = "bleed_ceoccur_v2", ref = "Y/N/NK" }
+if = { bleed_ceoccur_v2 = { "!=" = 3 } }
+occurrence_period = 
+#description = "Bleeding (Haemorrhage)"
+
+[[observation]]
+name = "cough"
+date = { ref = "admissionDateHierarchy" }
+phase = "admission"
+is_present = { if field = "cough_ceoccur_v2", ref = "Y/N/NK" }
```
Comment (abhidg) [inline isaric/parsers/rapid_observations.toml:31]: Should be `is_present = { field = "cough_ceoccur_v2", ref = "Y/N/NK" }`. Few other instances of this

File: isaric/parsers/rapid_observations.toml [hunk 3]
```diff
@@ -0,0 +1,782 @@
...
+is_present = { if field = "symptoms_cough", values = { 1 = true, 0 = false } }
+if = { symptoms_cough = { "!=" = "" } }
+occurrence_period = 
+#description = "Cough"
+
+[[observation]]
+name = "avpu"
+date = { ref = "admissionDateHierarchy" }
+phase = "study"
+value_char = { field = "avpu_vsorres", values = { 1 = "Alert", 2 = "Verbal", 3 = "Pain", 4 = "Unresponsive" }}
```
Comment (abhidg) [inline isaric/parsers/rapid_observations.toml:58]: Should these two avpu be combined? Unless they are a different phase or occurrence_period

File: isaric/parsers/rapid_observations.toml [hunk 4]
```diff
@@ -0,0 +1,782 @@
...
+date = { ref = "admissionDateHierarchy" }
+context = { field = "oxy_vsorresu", description = "Oxygen saturation on:", values = { 1 = ["Room air"], 2 = ["Oxygen therapy"]} }
+value_num = { field = "oxy_vsorres" }
+if = { oxy_vsyn = 1 }
+
+[[observation]]
+name = "oxygen_saturation_percent"
+phase = "admission"
+date = { ref = "admissionDateHierarchy" }
+context = { field = "daily_oxy_vsorresu", description = "Oxygen saturation on:", values = { 1 = ["Room air"], 2 = ["Oxygen therapy"]} }
```
Comment (abhidg) [inline isaric/parsers/rapid_observations.toml:80]: Same here, might be better to use firstNotNull for these

File: isaric/parsers/rapid_observations.toml [hunk 5]
```diff
@@ -0,0 +1,782 @@
...
+date = { ref = "followupDateHierarchy" }
+is_present = { if field = "flw_chest_pains", values = { 1 = true, 0 = false } }
+if = { flw_chest_pains = { "!=" = ""} }
+occurrence_period =
+
+[[observation]]
+name = "chest_pain"
+phase = "followup"
+date = { ref = "followupDateHierarchy" }
+is_present = { if field = "flw2_chest_pains", values = { 1 = true, 0 = false } }
```
Comment (abhidg) [inline isaric/parsers/rapid_observations.toml:255]: Can be combined with previous block

File: isaric/parsers/rapid_observations.toml [hunk 6]
```diff
@@ -0,0 +1,782 @@
...
+date = { ref = "followupDateHierarchy" }
+is_present = { if field = "flw_eq5d_mb_5l_uk_eng", values = { 2 = "I had slight problems in walking about", 3 = "I had moderate problems in walking about", 4 = "I had severe problems in walking about", 5 = "I was unable to walk about" } }
+if = { flw_eq5d_mb_5l_uk_eng = { "!=" = 1 } }
+occurrence_period = 
+
+[[observation]]
+name = "inability_walk"
+phase = "followup"
+date = { ref = "followupDateHierarchy" }
+is_present = { if field = "flw_eq5d_mb_5l_uk_eng_2", values = { 2 = "I had slight problems in walking about", 3 = "I had moderate problems in walking about", 4 = "I had severe problems in walking about", 5 = "I was unable to walk about" } }
```
Comment (abhidg) [inline isaric/parsers/rapid_observations.toml:663]: is_present is boolean

**Review by abhidg — COMMENTED** [review]
Thanks! some syntax errors, otherwise fine.

There are a few observations with the same name / phase / date but two different blocks (usually in phase = followup). Are these with different occurrence periods? If not, then they should be combined into one.

**Comment by ekamau** [comment]
Thanks. I will try that, later, when I have the bandwidth to get frustrated
with git conflicts ..

File: isaric/parsers/rapid_observations.toml (hunk 1 above)
Comment (ekamau) [reply isaric/parsers/rapid_observations.toml:24]: I will comment out the 'occurrence period'. I don't have values for these .. were they to be calculated automatically, based on two dates? or a selection from options? - to be discussed.

File: isaric/parsers/rapid_observations.toml (hunk 2 above)
Comment (ekamau) [reply isaric/parsers/rapid_observations.toml:31]: to be corrected.

File: isaric/parsers/rapid_observations.toml (hunk 3 above)
Comment (ekamau) [reply isaric/parsers/rapid_observations.toml:58]: there are mapping to two different fields.. one has 'daily_', which I imagine is a daily recording. We can discuss this further - whether to combine into one, or assume they have different occurrence period ..

File: isaric/parsers/rapid_observations.toml (hunk 5 above)
Comment (ekamau) [reply isaric/parsers/rapid_observations.toml:255]: we can discuss this ..

File: isaric/parsers/rapid_observations.toml (hunk 4 above)
Comment (ekamau) [reply isaric/parsers/rapid_observations.toml:80]: we can discuss the format for this. Now I really miss json. It was much easier to write come to think of it.

**Comment by ekamau** [comment]
Thanks for the reviews. Happy to discuss further and make changes.

**Pull request merged**